hw_lesson16.py

```diff
@@ -0,0 +1,203 @@
+"""
+Task 1
+Method overloading.
+
+Create a base class named Animal with a method called talk and then create two subclasses: Dog and Cat,
+and make their own implementation of the method talk be different.
+For instance, Dog’s can be to print ‘woof, woof’, while Cat’s can be to print ‘meow’.
+
+Also, create a simple generic function, which takes as input instance of a Cat or Dog classes
+and performs talk method on input parameter.
+"""
+
+
+class Animal:
+    def __init__(self, name):
+        self.name = name
+
+    def talk(self):
+        raise NotImplementedError('Animal can\'t speak')
+
+    def __str__(self):
+        return f'{self.name} says {self.talk()}'
+
+
+class Dog(Animal):
+
+    def talk(self):
+        return 'woof, woof'
+
+
+class Cat(Animal):
+
+    def talk(self):
+        return 'meow'
+
+
+"""
+Task 2
+Library
+
+Write a class structure that implements a library. Classes:
+
+1) Library - name, books = [], authors = []
+
+2) Book - name, year, author (author must be an instance of Author class)
+
+3) Author - name, country, birthday, books = []
+
+Library class
+
+Methods:
+
+- new_book(name: str, year: int, author: Author) - returns an instance of Book class
+and adds the book to the books list for the current library.
+
+- group_by_author(author: Author) - returns a list of all books grouped by the specified author
+
+- group_by_year(year: int) - returns a list of all the books grouped by the specified year
+
+All 3 classes must have a readable __repr__ and __str__ methods.
+
+Also, the book class should have a class variable which holds the amount of all existing books
+"""
+
+
+class Author:
+
+    def __init__(self, name, country, birthday, books=None):
+        if books is None:
+            books = []
+        self.name = name
+        self.country = country
+        self.birthday = birthday
+        self.books = books
+
+    def __repr__(self):
+        return f'Author: {self.name}'
+
+    def __str__(self):
+        return f'{self.name} has {self.books}. Author country - {self.country}'
+
+
+class Book:
+    count_books = 0
+
+    def __init__(self, name, year, author):
+        self.name = name
+        self.year = year
+        self.author = author
+        Book.count_books += 1
+
+    def __repr__(self):
+        return f'Book: {self.name}'
+
+    def __str__(self):
+        return f'{self.name} was written in {self.year} by {self.author}'
+
+
+class Library:
+
+    def __init__(self, name):
+        self.name = name
+        self.books = []
+        self.authors = []
+
+    def __repr__(self) -> str:
+        return f'{self.name}'
+
+    def __str__(self) -> str:
+        return f'Library "{self.name}" consist with {len(self.books)} books'
+
+    def new_book(self, name, year, author):
+        """ returns an instance of Book class and adds the book to the books list for the current library"""
+        book = Book(name, year, author)
+        self.books.append(book)
+        author.books.append(book.name)
```

Comment:
Have you missed it, haven't you?
```suggestion
        author.books.append(book.name)
        self.author.append(author)  # <<----
```

Reply:
Yes, i have mised this self.authors.append(author)
Corrected.